Source/Evolve/WarpXEvolve.cpp

```diff
@@ -237,7 +244,8 @@ WarpX::Evolve (int numsteps)
         // TODO: move out
         if (evolve_scheme == EvolveScheme::Explicit) {
             // At the end of last step, push p by 0.5*dt to synchronize
-            if (cur_time + dt[0] >= stop_time - 1.e-3*dt[0] || step == numsteps_max-1) {
+            if ((cur_time + dt[0] >= stop_time - 1.e-3*dt[0] || step == numsteps_max-1) &&
+                !synchronize_velocity_for_diagnostics) {
```

Comment:
Can you please add a helper function that returns true/false for `if the last step` query of the first part of this `if` and then simplify this `if`?

Also, you can reuse the logic for the `if` below.

Reply:
Is this better?